src/com/amazon/ion/impl/_Private_IonWriterBase.java

```diff
@@ -360,96 +354,128 @@ public void writeValue(IonReader reader) throws IOException
-                if (_debug_on) System.out.print(")");
-                break;
-            default:
-                throw new IllegalStateException("Unknown value type: " + type);
+            if (type == null) {
+                // There are no more values at this level. If we're at the starting level, we're done.
+                if (getDepth() == startingDepth) {
+                    break;
+                }
+                // Otherwise, step out once and then try to move forward again.
+                reader.stepOut();
+                stepOut();
+                continue;
             }
-        }
-    }
 
-    private void writeContainerRecursively(IonType type, IonReader reader)
-        throws IOException
-    {
-        stepIn(type);
-        reader.stepIn();
-        while ((type = reader.next()) != null)
-        {
-            writeValueRecursively(type, reader);
+            // We found a value. Write out its field ID and annotations, if any.
+            write_value_field_name_helper(reader);
+            write_value_annotations_helper(reader);
+
+            if (reader.isNullValue()) {
+                this.writeNull(type);
+                continue;
+            }
+
+            switch (type) {
+                case NULL:
+                    // The isNullValue() check above will handle this.
+                    throw new IllegalStateException("isNullValue() was false but IonType was NULL.");
+                case BOOL:
+                    writeBool(reader.booleanValue());
+                    break;
+                case INT:
+                    writeInt(reader.bigIntegerValue());
+                    break;
+                case FLOAT:
+                    writeFloat(reader.doubleValue());
+                    break;
+                case DECIMAL:
+                    writeDecimal(reader.decimalValue());
+                    break;
+                case TIMESTAMP:
+                    writeTimestamp(reader.timestampValue());
+                    break;
+                case STRING:
+                    writeString(reader.stringValue());
+                    break;
+                case SYMBOL:
+                    writeSymbolToken(reader.symbolValue());
+                    break;
+                case BLOB:
+                    writeBlob(reader.newBytes());
+                    break;
+                case CLOB:
+                    writeClob(reader.newBytes());
+                    break;
+                case STRUCT:
+                    reader.stepIn();
+                    stepIn(IonType.STRUCT);
+                    break;
+                case LIST:
+                    reader.stepIn();
+                    stepIn(IonType.LIST);
+                    break;
+                case SEXP:
+                    reader.stepIn();
+                    stepIn(IonType.SEXP);
+                    break;
+                default:
+                    throw new IllegalStateException("Unknown value type: " + type);
```

Comment:
Your mileage may vary here, but what do you think of this expression?
```java
switch (type) {
    case      NULL: // The isNullValue() check above will handle this.
        throw new IllegalStateException("isNullValue() was false but IonType was NULL.");
    
    case      BOOL: writeBool(reader.booleanValue());        break;
    case       INT: writeInt(reader.bigIntegerValue());      break;
    case     FLOAT: writeFloat(reader.doubleValue());        break;
    case   DECIMAL: writeDecimal(reader.decimalValue());     break;
    case TIMESTAMP: writeTimestamp(reader.timestampValue()); break;
    case    STRING: writeString(reader.stringValue());       break;
    case    SYMBOL: writeSymbolToken(reader.symbolValue());  break;
    case      BLOB: writeBlob(reader.newBytes());            break;
    case      CLOB: writeClob(reader.newBytes());            break;
    case    STRUCT: // Intentional fallthrough
    case      LIST: // Intentional fallthrough
    case      SEXP: reader.stepIn(); stepIn(type);           break;

    default:
        throw new IllegalStateException("Unexpected value type: " + type);
}
```

Reply:
I adopted the fallthrough for the container types, thanks!